viewer/core/src/utilities/worldToViewport.ts

```diff
@@ -4,32 +4,76 @@
 
 import * as THREE from 'three';
 
-// Should be somewhere else??
-interface Position2D {
-  x: number;
-  y: number;
-}
+const worldToViewportVars = {
+  renderSize: new THREE.Vector2(),
+  position: new THREE.Vector3()
+};
 
-export function from3DPositionToRelativeViewportCoordinates(
+/**
+ * Maps from 3D world coordinates to normalized screen coordinates
+ * relative to the canvas being rendered to. X and Y will
+ * be in range [0, 1]. Z is in range [-1, 1] if the coordinate
+ * is inside the camera near and far planes.
+ * @param renderer    Renderer used to render the "world"
+ * @param camera      Camera used to project point from 3D to NDC coordinates
+ * @param position3D  World position in 3D
+ * @param out         Optionally pre-allocated THREE.Vector3
+ * @returns           Relative screen coordinates in X, Y and Z in range [-1, 1] if point
+ * is within near/far of camera.
+ */
+export function worldToNormalizedViewportCoordinates(
+  renderer: THREE.WebGLRenderer,
   camera: THREE.PerspectiveCamera,
-  position3D: THREE.Vector3
+  position3D: THREE.Vector3,
+  out: THREE.Vector3 = new THREE.Vector3()
 ): THREE.Vector3 {
-  const relativePosition = position3D.clone().project(camera);
-  relativePosition.y = -relativePosition.y;
-  relativePosition.addScalar(1).multiplyScalar(0.5); // Is it better to have a multiply with a 0.5, -0.5 vector?
-  return relativePosition;
+  const { renderSize, position } = worldToViewportVars;
+  const canvas = renderer.domElement;
+  const pixelRatio = 1; // renderer.getPixelRatio();
+  renderer.getSize(renderSize);
+
+  // map to normalized device coordinate (NDC) space
+  position.copy(position3D);
+  position.project(camera);
```

Comment:
I guess the project method does perspective divide?

Reply:
It does, ref https://github.com/mrdoob/three.js/blob/e62b253081438c030d6af1ee3c3346a89124f277/src/math/Vector3.js#L301 + https://github.com/mrdoob/three.js/blob/e62b253081438c030d6af1ee3c3346a89124f277/src/math/Vector3.js#L262